### PR TITLE
[release/8.0] Metrics Feature Switch

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/89880

Backport of #91767 to release/8.0

/cc @tarekgh

## Customer Impact

The introduction of a feature switch enables the option to disable the metrics feature, aiding in trimming and AOT scenarios. This modification addresses and resolves the performance regression observed in Xamarin app startup scenarios. For additional information, please refer to the following issue: https://github.com/dotnet/runtime/issues/89880.

## Testing

Successfully ran all regressions test and added a new test covering testing when the new switch is enabled. 

## Risk
We haven't made any alterations or adjustments to the code or logic when the new switch is left disabled. Therefore, any modifications will only become apparent when the switch is activated. The risk of this change adversely affecting any existing, functional functionality is exceedingly low.

**IMPORTANT**: If this backport is for a servicing release, please verify that:

- The PR target branch is `release/X.0-staging`, not `release/X.0`.

- If the change touches code that ships in a NuGet package, you have added the necessary [package authoring](https://github.com/dotnet/runtime/blob/main/docs/project/library-servicing.md) and gotten it explicitly reviewed.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @noahfalk @eerhardt

**Comment by tarekgh** [comment]
@ericstj please have a look and approve it if you don't see any concerns with it. Thanks!

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/TestNotSupported.cs
Comment (eerhardt) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/TestNotSupported.cs:47]:
```suggestion
                bool isSupported = bool.Parse(val);

                Meter meter = new Meter("IsSupportedTest");
                Counter<long> counter = meter.CreateCounter<long>("counter");
                bool instrumentsPublished = false;
                bool instrumentCompleted = false;
                long counterValue = 100;

                using (MeterListener listener = new MeterListener
                {
                    InstrumentPublished = (instruments, theListener) => instrumentsPublished = true,
                    MeasurementsCompleted = (instruments, state) => instrumentCompleted = true
                })
                {
                    listener.EnableMeasurementEvents(counter, null);
                    listener.SetMeasurementEventCallback<long>((inst, measurement, tags, state) => counterValue = measurement);
                    listener.Start();

                    Assert.Equal(isSupported, counter.Enabled);

                    counter.Add(20);
                }
                meter.Dispose();

                Assert.Equal(isSupported, instrumentsPublished);
                Assert.Equal(isSupported, instrumentCompleted);
                Assert.Equal(isSupported ? 20 : 100, counterValue);
```

(nit) A few newlines are in `main`, but were lost here in the backport. You might want to make sure the files are identical in case a merge needs to happen.

**Review by eerhardt — APPROVED** [review]

**Review by ericstj — APPROVED** [review]
This LGTM and I approve for RC2.

One ask is to double check that this will integrate well by staging the changes you want to make in the SDK and confirming the right default / feature enabled behavior.  I know I've mixed up defaults or true/false values more than once myself.

**Comment by ericstj** [comment]
@artl93 this is ready for your approval as well.

**Comment by tarekgh** [comment]
> One ask is to double check that this will integrate well by staging the changes you want to make in the SDK and confirming the right default / feature enabled behavior. I know I've mixed up defaults or true/false values more than once myself.

I am currently doing some more tests. Please hold merging till I finish. nothing alarming so far :-)

**Comment by tarekgh** [comment]
@ericstj I have done more manual testing building and running as AOT. I confirmed when the switch is not defined or it is defined with `true` value, I see the metrics functionality is enabled and work as expected. When defining the switch with `false` value, I am seeing the new behavior which disables metrics functionality. I think we are good to go. I haven't done the SDK changes yet which I'll do more testing when having these changes.

**Comment by artl93** [comment]
> @ericstj I have done more manual testing building and running as AOT. I confirmed when the switch is not defined or it is defined with `true` value, I see the metrics functionality is enabled and work as expected. When defining the switch with `false` value, I am seeing the new behavior which disables metrics functionality. I think we are good to go. I haven't done the SDK changes yet which I'll do more testing when having it.

For my education, what do you mean by, "When having it"? Do you mean after merge and the SDK is produced?

**Comment by ericstj** [comment]
This change in dotnet/runtime will require corresponding change in the SDK targets to expose this new feature to the build.   Those changes will look similar to these:
https://github.com/search?q=repo%3Adotnet%2Fsdk+EventSourceSupport+language%3AXML&type=code&l=XML

**Review by artl93 — APPROVED** [review]
M2 approved. Thanks!

**Pull request merged**
